modules/mpa/decode.c

```diff
@@ -0,0 +1,182 @@
+/**
+ * @file mpa/decode.c mpa Decode
+ *
+ * Copyright (C) 2016 Symonics GmbH
+ */
+
+#include <re.h>
+#include <baresip.h>
+#include <mpg123.h>
+#include <speex/speex_resampler.h>
+#include <string.h>
+#include "mpa.h"
+
+#undef DEBUG
+
+struct audec_state {
+	mpg123_handle *dec;
+	SpeexResamplerState *resampler;
+	int channels;
+};
+
+
+static void destructor(void *arg)
+{
+	struct audec_state *ads = arg;
+
+	mpg123_close(ads->dec);
+	mpg123_delete(ads->dec);
+#ifdef DEBUG
+	debug("mpa: decoder destroyed\n");
+#endif
+}
+
+
+int mpa_decode_update(struct audec_state **adsp, const struct aucodec *ac,
+		       const char *fmtp)
+{
+	struct audec_state *ads;
+	int mpaerr, err=0;
+
+	if (!adsp || !ac || !ac->ch)
+		return EINVAL;
+
+	ads = *adsp;
+
+#ifdef DEBUG
+	debug("mpa: decoder created %s\n",fmtp);
+#endif
+
+	if (ads)
+		mem_deref(ads);
+		
+	ads = mem_zalloc(sizeof(*ads), destructor);
+	if (!ads)
+		return ENOMEM;
+	ads->channels = 0;
+	ads->resampler = NULL;
+
+	ads->dec = mpg123_new(NULL,&mpaerr);
+	if (!ads->dec) {
+		error("mpa: decoder create: %s\n", mpg123_plain_strerror(mpaerr));
+		err = ENOMEM;
+		goto out;
+	}
+
+#ifdef DEBUG
+	mpaerr = mpg123_param(ads->dec, MPG123_VERBOSE, 4, 4.);
+#else
+	mpaerr = mpg123_param(ads->dec, MPG123_VERBOSE, 0, 0.);
+#endif
+	if(mpaerr != MPG123_OK) {
+		error("MPA libmpg123 param error %s", mpg123_plain_strerror(mpaerr));
+		err = EINVAL;
+		goto out;
+	}
+
+
+	mpaerr = mpg123_format_all(ads->dec);
+	if(mpaerr != MPG123_OK) {
+		error("MPA libmpg123 format error %s", mpg123_plain_strerror(mpaerr));
+		err = EINVAL;
+		goto out;
+	}
+
+	mpaerr = mpg123_open_feed(ads->dec);
+	if(mpaerr != MPG123_OK) {
+		error("MPA libmpg123 open feed error %s", mpg123_plain_strerror(mpaerr));
+		err = EINVAL;
+		goto out;
+	}
+
+
+ out:
+	if (err)
+		mem_deref(ads);
+	else
+		*adsp = ads;
+
+	return err;
+}
+
+
+int mpa_decode_frm(struct audec_state *ads, int16_t *sampv, size_t *sampc,
+		    const uint8_t *buf, size_t len)
+{
+	int mpaerr, channels, encoding, i;
+	long samplerate,res;
+	size_t n;
+	uint32_t header;
+	static int16_t ds[2304];
```

Comment:
the ds buffer is static, meaning multiple instances of mpa will not work. You should either declare
it on the stack or move it to the audec_state.